enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,225 @@
+therefore custom certificates will have to be validated against CAs in the RHEL Client trust store. 
+
+Each certificate gets examined in order to determine if it's purely wildcard.
+A certificate is considered to be purely wildcard only if there are no FQDN Entries found.
+Certificate will be considered as a wildcard only if there are no FQDN Entries found.
+The FQDN Address is searched in:
+- Certificate Subject Alternative Name (SAN)
+- names configuration value.
+
+when a Certificate is found to be a wildcard only, Microshift will not be able to find the real FQDN, node IP Address will be placed inside its kubeconfig's `server` section.
+this kubeconfig should be treated as an example only and its `server` section value should be changed to the real FQDN.
+
+Certain values that configured in `names` field can cause  internal API communication issues  ie:127.0.0.1,localhost therefore we must not allow them.
+
+### Workflow Description
+
+#### Default API Certs
+1. By default, when there is no namedCertificates configuration the behaviour will remain the same.
+
+#### when custom namedCertificates configured
+1. Device Administrator copies the certificated to MicroShift host,files should be accessible by root only.
+1. Device Administrator configures additonal CAs in the RHEL Client trust store on the client system.
+1. Device Administrator configures `namedCertificates` in the Microshift configuration yaml file (/etc/microshift/config.yaml).
+1. Device Administrator start/restarts MicroShift
+1. Check and validate the certificates see [Failure Modes](#failure-modes).
+1. Microshift passes the certificates to the tls-sni-cert-key as apiserver command line option preceding all the other certificates.
+1. kube-apiserver picks up the certificates and start serving them on the configured SNI.
+
+### Topology Considerations
+#### Hypershift / Hosted Control Planes
+N/A
+
+#### Standalone Clusters
+N/A
+
+#### Single-node Deployments or MicroShift
+Enhancement is solely intended for MicroShift.
+
+### Implementation Details/Notes/Constraints
+The certs will be prepended to the []configv1.NamedCertificate list before the api server is started. (it will be added to `-tls-sni-cert-key` flag)
+
+This certificate paths configuration and names will be prepended into the kube-apiserver `tls-sni-cert-key` command line flag.
+When the same SNI appears in the SAN part of the provided certs, this certificate will take precedence over the `default` external-signer. [ref](https://github.com/kubernetes/kubernetes/blob/98358b8ce11b0c1878ae7aa1482668cb7a0b0e23/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go#L38)
+
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+* User provide certificate and it expired after some time.
+  kube-api server will continue serving with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
```

Comment:
Is this a mitigation that would be recommended in practice to real users?

Reply:
well, since the certificate is issued by the the actual client,we must provide some means to overcome its expiration.